File: crates/lib/superdev-core/assets/templates/web-react-android-ios-native/libs/native-debug-server-android/src/main/kotlin/com/nativedebugserver/LogBuffer.kt
// Thread-safe circular log buffer for debug log entries.

package com.nativedebugserver

import java.time.Instant
import java.time.format.DateTimeFormatter

/**
 * A single log entry.
 */
data class LogEntry(
    val ts: Instant,
    val level: Level,
    val tag: String,
    val message: String
) {
    enum class Level { DEBUG, INFO, WARN, ERROR }

    fun toMap(): Map<String, Any> = mapOf(
        "ts" to DateTimeFormatter.ISO_INSTANT.format(ts),
        "level" to level.name.lowercase(),
        "tag" to tag,
        "message" to message
    )
}

/**
 * Thread-safe circular buffer for log entries.
 */
class LogBuffer(private val capacity: Int = 1000) {
    private val entries = mutableListOf<LogEntry>()
    private val lock = Any()

    fun append(level: LogEntry.Level, tag: String, message: String) {
        val entry = LogEntry(
            ts = Instant.now(),
            level = level,
            tag = tag,
            message = message
        )
        synchronized(lock) {
            entries.add(entry)
            if (entries.size > capacity) {
                entries.removeAt(0)
            }
        }
    }

    fun getEntries(since: Instant? = null): List<LogEntry> {
        synchronized(lock) {
            return if (since != null) {
                entries.filter { it.ts.isAfter(since) }
            } else {
                entries.toList()
            }
        }
    }

    fun clear() {
        synchronized(lock) {
            entries.clear()
        }
    }

    val count: Int
        get() = synchronized(lock) { entries.size }
}

File: crates/lib/superdev-core/assets/templates/web-react-android-ios-native/libs/native-debug-server-ios/Sources/NativeDebugServer/LogBuffer.swift
// Thread-safe circular log buffer.

import Foundation

/// A single log entry.
public struct LogEntry: Sendable {
    public enum Level: String, Sendable {
        case debug, info, warn, error
    }

    public let ts: Date
    public let level: Level
    public let tag: String
    public let message: String

    func toDict() -> [String: Any] {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return [
            "ts": formatter.string(from: ts),
            "level": level.rawValue,
            "tag": tag,
            "message": message
        ]
    }
}

/// Thread-safe circular buffer for log entries.
public final class LogBuffer: Sendable {
    private let queue = DispatchQueue(label: "com.nativedebugserver.logbuffer")
    private nonisolated(unsafe) var entries: [LogEntry] = []
    private let capacity: Int

    public init(capacity: Int = 1000) {
        self.capacity = capacity
    }

    /// Append a new log entry.
    public func append(level: LogEntry.Level, tag: String, message: String) {
        let entry = LogEntry(ts: Date(), level: level, tag: tag, message: message)
        queue.sync {
            entries.append(entry)
            if entries.count > capacity {
                entries.removeFirst(entries.count - capacity)
            }
        }
    }

    /// Get all entries, optionally filtered by timestamp.
    public func getEntries(since: Date? = nil) -> [LogEntry] {
        queue.sync {
            guard let since = since else { return entries }
            return entries.filter { $0.ts > since }
        }
    }

    /// Clear all entries.
    public func clear() {
        queue.sync {
            entries.removeAll()
        }
    }

    /// Current count.
    public var count: Int {
        queue.sync { entries.count }
    }
}
